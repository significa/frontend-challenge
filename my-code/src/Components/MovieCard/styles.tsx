import styled from 'styled-components';

export const MovieCardContainer = styled.div`
    width: fit-content;
    height: fit-content;
    position: relative;
`;

export const MoviePoster = styled.img`
    position: relative;
    width: 180px;
    height: 240px;
    border-radius: 3px;

`;

export const MovieDetails = styled.div`
    position: absolute;
    top: 0;
    right: 0;
    width: 100%;
    height: 100%;
    border-radius: 3px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    opacity: 0;
    transition: all 0.2s ease-in-out;
    z-index: 1;

    :hover {
        opacity: 1;

        & + ${MoviePoster} {
            filter: brightness(0.5);
        }
    }

`;

export const FavoriteIcon = styled.div`
    width: 100%;
    display: flex;
    justify-content: flex-end;
    padding: .5rem;
    cursor: pointer;
`;

export const MovieInfo = styled.div`
    padding: .5rem;
    cursor: pointer;
`;
