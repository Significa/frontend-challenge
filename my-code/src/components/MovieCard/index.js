import React from 'react'
import PropTypes from 'prop-types'

import Heart from '../../assets/icon-heart.svg'
import HeartFilled from '../../assets/icon-heart-filled.svg'

import {
  Wrapper,
  CoverLink,
  Poster,
  FavoriteButtonWrapper,
  Title,
  Subtitle,
  Content
} from './style'

export default function MovieCard({
  id,
  title,
  poster,
  year,
  favorite,
  link,
  toggleFavorite
}) {
  return (
    <Wrapper>
      {poster && <Poster poster={poster} />}
      <CoverLink to={link} aria-label={title} />
      <FavoriteButtonWrapper
        icon={favorite ? HeartFilled : Heart}
        visible={favorite || !poster}
        onClick={() => toggleFavorite(id, !favorite)}
        label={`${favorite ? 'Unfavorite' : 'Favorite'} ${title}`}
      />
      <Content>
        <Title>{title}</Title>
        <Subtitle>{year}</Subtitle>
      </Content>
    </Wrapper>
  )
}

MovieCard.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  poster: PropTypes.string,
  year: PropTypes.string.isRequired,
  favorite: PropTypes.bool.isRequired,
  link: PropTypes.string.isRequired,
  toggleFavorite: PropTypes.func.isRequired
}

MovieCard.defaultProps = {
  poster: undefined
}
