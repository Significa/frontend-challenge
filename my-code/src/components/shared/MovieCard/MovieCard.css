@import '../../../style-vars.css';

.Wrapper {
  position: relative;
  font-family: var(--font-family);
  border-radius: var(--border-radius-1);
  overflow: hidden;
  cursor: pointer;

  &:hover .Info {
    display: block;
  }
}

.Poster {
  display: block;
  object-fit: cover;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
}

.PosterError {
  border-radius: 5px;
  border: 2px solid var(--color-mid-grey);
}

.Info {
  display: block;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: var(--color-grey-opacity);
  
  
  @media (--tablet) {
    display: none;
  }
}

.Content {
  box-sizing: border-box;
  padding: var(--spacing-4);
  position: relative;
  height: 100%;
  display: flex;
  align-self: flex-end;
  color: var(--color-white);
  text-decoration: none;
  justify-content: flex-end;
  flex-direction: column;
}

.Title {
  margin: 0;
  font-weight: var(--font-weigth-medium);
  font-size: var(--font-size-2);
  letter-spacing: var(--letter-spacing-1);
  margin-bottom: var(--spacing-1);
}

.LikeButton {
  background: none;
  border: none;
  appearance: none;
  outline: none;

  padding: 0;
  position: absolute;
  right: var(--spacing-4);
  top: var(--spacing-4);
}
