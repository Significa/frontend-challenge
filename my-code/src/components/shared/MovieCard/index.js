// @flow
export { default } from './MovieCard';
