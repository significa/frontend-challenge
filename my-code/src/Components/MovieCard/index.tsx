import { useNavigate } from 'react-router-dom';
import { useMovies } from '../../Contexts/MoviesData';
import { FavouriteIcon, MovieCardContainer, MovieDetails, MovieInfo, MoviePoster } from './styles';
import fullHeartIcon from '/icons/icon-heart-full.svg';
import whiteHeartIcon from '/icons/icon-heart-white.svg';

interface MovieCardType {
    poster: string
    title: string
    year: string
    imdbID: string
}

export function MovieCard({ imdbID, poster, title, year }: MovieCardType) {

    const navigate = useNavigate()
    const { isFavourite, addToFavourites } = useMovies()

    function handleClick() {
        navigate(`/movie/${imdbID}`)
    }

    return (
        <MovieCardContainer>
            <MovieDetails>
                <MovieInfo onClick={handleClick}>
                    <h2>{title}</h2>
                    <p>{year.length === 5 ? `${year} OnGoing` : year}</p>
                </MovieInfo>
            </MovieDetails>
            <FavouriteIcon isFavourite={isFavourite(imdbID)}>
                <img
                    onClick={() => addToFavourites(imdbID)}
                    src={isFavourite(imdbID) ? fullHeartIcon : whiteHeartIcon}
                    alt="Add to favourites"
                />
            </FavouriteIcon>
            <MoviePoster src={poster != 'N/A' ? poster : 'https://image-placeholder.com/images/actual-size/180x240.png'} alt={title} />
        </MovieCardContainer>
    );
}