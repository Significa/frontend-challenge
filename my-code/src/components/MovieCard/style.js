import sc from 'styled-components'
import { Link } from 'react-router-dom'
import { theme } from 'styled-tools'

import { text3, text4 } from '../../styles/text'
import IconButton from '../IconButton'

export const Wrapper = sc.div`
  width: 100%;
  text-decoration: none;
  position: relative;
  border-radius: 3px;
  overflow: hidden;
  color: ${theme('greyWhite')};

  &:focus {
    outline: none;
  }

  /* Aspect ratio */
  &:before {
    content: '';
    display: block;
    padding-top: 133.3333333%;
  }
`

export const CoverLink = sc(Link)`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`

export const Poster = sc.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: ${theme('greyMain')};
  background-image: ${({ poster }) => `url(${poster})`};
`

export const FavoriteButtonWrapper = sc(IconButton)`
  height: auto;
  position: absolute;
  z-index: 1;
  top: 12px;
  right: 12px;
  color: ${theme('greyWhite')};
  will-change: opacity;
  transition:
    opacity
    ${theme('transitionTiming')}
    ${theme('transitionDuration')}
  ;
  filter: drop-shadow(0 0 2px rgba(0, 0, 0, .2));
  opacity: ${({ visible = true }) => (visible ? 1 : 0)};

  ${Poster} ~ ${CoverLink}:hover ~ &,
  ${Poster} ~ ${CoverLink}:focus ~ &,
  ${Poster} ~ &:hover,
  ${Poster} ~ &:focus {
    opacity: 1;
  }
`

export const Title = sc.span`
  ${text3}
`

export const Subtitle = sc.span`
  ${text4}
`

export const Content = sc.div`
  pointer-events: none;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: 12px;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  background: ${theme('scrimColor')};
  will-change: opacity;
  transition:
    opacity
    ${theme('transitionTiming')}
    ${theme('transitionDuration')}
  ;

  ${Poster} ~ & {
    opacity: 0;
  }

  ${Poster} ~ ${CoverLink}:hover ~ &,
  ${Poster} ~ ${CoverLink}:focus ~ &,
  ${Poster} ~ ${FavoriteButtonWrapper}:hover ~ &,
  ${Poster} ~ ${FavoriteButtonWrapper}:focus ~ & {
    opacity: 1;
  }
`
