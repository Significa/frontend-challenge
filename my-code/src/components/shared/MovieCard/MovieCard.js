// @flow

import React, { useState } from 'react';
import Link from 'next/link';
import classNames from 'classnames';
import heartIconFull from '../../../icons/icon-heart-full.svg';
import heartIconWhite from '../../../icons/icon-heart-white.svg';
import styles from './MovieCard.css';
import { useMoviesState } from '../../context/MoviesContext';
import imgFallback from '../../../icons/logo-rotten-tomatoes.svg';

type Props = {
  liked: boolean,
  title: string,
  id: string,
  poster: string,
  year: string,
  className?: string
};

export default function MovieCard(props: Props) {
  const { title, year, poster, id, className } = props;
  const [image, setImage] = useState(poster);
  const { state, dispatch } = useMoviesState();
  const isLiked = state.includes(id);

  const handleLike = () => {
    const isLiked = state.includes(id);
    const actionType = isLiked ? 'UNLIKE' : 'LIKE';

    dispatch({
      type: actionType,
      payload: {
        id
      }
    });
  };

  const handleImageError = () => {
    setImage(imgFallback);
  };

  return (
    <div className={classNames(styles.Wrapper, className)}>
      <img onError={handleImageError} alt={`${title} Poster`} className={styles.Poster} src={image} />
      <div className={styles.Info}>
        <Link shallow href="/movie/[id]" as={`/movie/${id}`}>
          <a className={styles.Content}>
            <h2 className={styles.Title}>{title}</h2>
            <span>{year}</span>
          </a>
        </Link>
        <button className={styles.LikeButton}>
          <img
            onClick={handleLike}
            alt={isLiked ? 'Liked icon' : 'Like icon'}
            src={isLiked ? heartIconFull : heartIconWhite}
          />
        </button>
      </div>
    </div>
  );
}
