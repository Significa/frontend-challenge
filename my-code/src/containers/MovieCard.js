import { connect } from 'react-redux'

import { toggleFavorite, getIsFavorite } from '../state/modules/favorites'
import { getMovie } from '../state/modules/movies'

import MovieCard from '../components/MovieCard'

const mapStateToProps = ({ movies, favorites }, { id }) => ({
  ...getMovie(movies, id),
  favorite: getIsFavorite(favorites, id)
})

const mapDispatchToProps = { toggleFavorite }

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovieCard)
