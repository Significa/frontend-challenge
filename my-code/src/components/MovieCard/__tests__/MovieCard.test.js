import React from 'react'
import { shallow } from 'enzyme'

import MovieCard from '../'
import {
  Wrapper,
  CoverLink,
  Poster,
  FavoriteButtonWrapper,
  Title,
  Subtitle,
  Content
} from '../style'

function MockIcon() {
  return null
}

describe('MovieCard', () => {
  it('renders correctly', () => {
    const tree = shallow(
      <MovieCard
        id="tt1375666"
        title="The Movie Title"
        poster="https://poster.url"
        year="2000"
        favorite={false}
        link="/movie-link"
        toggleFavorite={() => {}}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders without poster correctly', () => {
    const tree = shallow(
      <MovieCard
        id="tt1375666"
        title="The Movie Title"
        year="2000"
        favorite={false}
        link="/movie-link"
        toggleFavorite={() => {}}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders favorite correctly', () => {
    const tree = shallow(
      <MovieCard
        id="tt1375666"
        title="The Movie Title"
        poster="https://poster.url"
        year="2000"
        favorite
        link="/movie-link"
        toggleFavorite={() => {}}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders favorite without poster correctly', () => {
    const tree = shallow(
      <MovieCard
        id="tt1375666"
        title="The Movie Title"
        year="2000"
        favorite
        link="/movie-link"
        toggleFavorite={() => {}}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('calls toggleFavorite on click', () => {
    const mockCallback = jest.fn()
    const tree = shallow(
      <MovieCard
        id="tt1375666"
        title="The Movie Title"
        year="2000"
        favorite
        link="/movie-link"
        toggleFavorite={mockCallback}
      />
    )

    tree.find(FavoriteButtonWrapper).simulate('click')
    expect(mockCallback.mock.calls.length).toBe(1)
    expect(mockCallback.mock.calls[0][0]).toBe('tt1375666')
    expect(mockCallback.mock.calls[0][1]).toBe(false)
  })

  describe('Wrapper', () => {
    it('renders correctly', () => {
      const tree = shallow(<Wrapper />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('CoverLink', () => {
    it('renders correctly', () => {
      const tree = shallow(<CoverLink to="some-url" />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Poster', () => {
    it('renders correctly', () => {
      const tree = shallow(<Poster poster="poster-url" />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('FavoriteButtonWrapper', () => {
    it('renders correctly', () => {
      const tree = shallow(
        <FavoriteButtonWrapper icon={MockIcon} onClick={() => {}} />
      )
      expect(tree).toMatchSnapshot()
    })

    it('renders non-visble correctly', () => {
      const tree = shallow(
        <FavoriteButtonWrapper
          icon={MockIcon}
          onClick={() => {}}
          visible={false}
        />
      )
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Title', () => {
    it('renders correctly', () => {
      const tree = shallow(<Title />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Subtitle', () => {
    it('renders correctly', () => {
      const tree = shallow(<Subtitle />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Content', () => {
    it('renders correctly', () => {
      const tree = shallow(<Content />)
      expect(tree).toMatchSnapshot()
    })
  })
})
