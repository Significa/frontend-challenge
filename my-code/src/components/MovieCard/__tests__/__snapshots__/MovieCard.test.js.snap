// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Content renders correctly 1`] = `
.c0 {
  pointer-events: none;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  padding: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  background: theme.scrimColor;
  will-change: opacity;
  -webkit-transition: opacity theme.transitionTiming theme.transitionDuration;
  transition: opacity theme.transitionTiming theme.transitionDuration;
}

.c2 ~ .c0 {
  opacity: 0;
}

.c2 ~ .c1:hover ~ .c0,
.c2 ~ .c1:focus ~ .c0,
.c2 ~ .FavoriteButtonWrapper-eTymEw:hover ~ .c0,
.c2 ~ .FavoriteButtonWrapper-eTymEw:focus ~ .c0 {
  opacity: 1;
}

<div
  className="c0"
/>
`;

exports[`MovieCard CoverLink renders correctly 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<Link
  className="c0"
  replace={false}
  to="some-url"
/>
`;

exports[`MovieCard FavoriteButtonWrapper renders correctly 1`] = `
.c0 {
  height: auto;
  position: absolute;
  z-index: 1;
  top: 12px;
  right: 12px;
  color: theme.greyWhite;
  will-change: opacity;
  -webkit-transition: opacity theme.transitionTiming theme.transitionDuration;
  transition: opacity theme.transitionTiming theme.transitionDuration;
  -webkit-filter: drop-shadow(0 0 2px rgba(0,0,0,.2));
  filter: drop-shadow(0 0 2px rgba(0,0,0,.2));
  opacity: 1;
}

.c2 ~ .c1:hover ~ .c0,
.c2 ~ .c1:focus ~ .c0,
.c2 ~ .c0:hover,
.c2 ~ .c0:focus {
  opacity: 1;
}

<IconButton
  className="c0"
  icon={[Function]}
  onClick={[Function]}
/>
`;

exports[`MovieCard FavoriteButtonWrapper renders non-visble correctly 1`] = `
.c0 {
  height: auto;
  position: absolute;
  z-index: 1;
  top: 12px;
  right: 12px;
  color: theme.greyWhite;
  will-change: opacity;
  -webkit-transition: opacity theme.transitionTiming theme.transitionDuration;
  transition: opacity theme.transitionTiming theme.transitionDuration;
  -webkit-filter: drop-shadow(0 0 2px rgba(0,0,0,.2));
  filter: drop-shadow(0 0 2px rgba(0,0,0,.2));
  opacity: 0;
}

.c2 ~ .c1:hover ~ .c0,
.c2 ~ .c1:focus ~ .c0,
.c2 ~ .c0:hover,
.c2 ~ .c0:focus {
  opacity: 1;
}

<IconButton
  className="c0"
  icon={[Function]}
  onClick={[Function]}
  visible={false}
/>
`;

exports[`MovieCard Poster renders correctly 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-position: center center;
  background-repeat: no-repeat;
  background-size: cover;
  background-color: theme.greyMain;
  background-image: url(poster-url);
}

<div
  className="c0"
  poster="poster-url"
/>
`;

exports[`MovieCard Subtitle renders correctly 1`] = `
.c0 {
  font-weight: 400;
  font-size: 16px;
  line-height: 1.5em;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
}

<span
  className="c0"
/>
`;

exports[`MovieCard Title renders correctly 1`] = `
.c0 {
  font-weight: 500;
  font-size: 20px;
  line-height: 1.4em;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
}

<span
  className="c0"
/>
`;

exports[`MovieCard Wrapper renders correctly 1`] = `
.c0 {
  width: 100%;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  border-radius: 3px;
  overflow: hidden;
  color: theme.greyWhite;
}

.c0:focus {
  outline: none;
}

.c0:before {
  content: '';
  display: block;
  padding-top: 133.3333333%;
}

<div
  className="c0"
/>
`;

exports[`MovieCard renders correctly 1`] = `
<Wrapper>
  <Poster
    poster="https://poster.url"
  />
  <CoverLink
    aria-label="The Movie Title"
    to="/movie-link"
  />
  <FavoriteButtonWrapper
    icon={[Function]}
    label="Favorite The Movie Title"
    onClick={[Function]}
    visible={false}
  />
  <Content>
    <Title>
      The Movie Title
    </Title>
    <Subtitle>
      2000
    </Subtitle>
  </Content>
</Wrapper>
`;

exports[`MovieCard renders favorite correctly 1`] = `
<Wrapper>
  <Poster
    poster="https://poster.url"
  />
  <CoverLink
    aria-label="The Movie Title"
    to="/movie-link"
  />
  <FavoriteButtonWrapper
    icon={[Function]}
    label="Unfavorite The Movie Title"
    onClick={[Function]}
    visible={true}
  />
  <Content>
    <Title>
      The Movie Title
    </Title>
    <Subtitle>
      2000
    </Subtitle>
  </Content>
</Wrapper>
`;

exports[`MovieCard renders favorite without poster correctly 1`] = `
<Wrapper>
  <CoverLink
    aria-label="The Movie Title"
    to="/movie-link"
  />
  <FavoriteButtonWrapper
    icon={[Function]}
    label="Unfavorite The Movie Title"
    onClick={[Function]}
    visible={true}
  />
  <Content>
    <Title>
      The Movie Title
    </Title>
    <Subtitle>
      2000
    </Subtitle>
  </Content>
</Wrapper>
`;

exports[`MovieCard renders without poster correctly 1`] = `
<Wrapper>
  <CoverLink
    aria-label="The Movie Title"
    to="/movie-link"
  />
  <FavoriteButtonWrapper
    icon={[Function]}
    label="Favorite The Movie Title"
    onClick={[Function]}
    visible={true}
  />
  <Content>
    <Title>
      The Movie Title
    </Title>
    <Subtitle>
      2000
    </Subtitle>
  </Content>
</Wrapper>
`;
